File: joshuaCarpentier/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.js
// Problem Statement #

// We are given an array containing ‘n’ objects. Each object, when created, was assigned a unique number from 1 to ‘n’ 
// based on their creation sequence. This means that the object 
// with sequence number ‘3’ was created just before the object with sequence number ‘4’.

// Write a function to sort the objects in-place on their creation sequence number in O(n) and
//  without any extra space. For simplicity, let’s assume we are passed an integer array containing 
//  only the sequence numbers, though each number is actually an object.


/**
 * Input: Unorder array
 * Output: Sorted array
 * 
 * while current index less then array length if 
 * current index not equal to index
 * 
 * next index  equals current index - 1
 * 
 * then inddex  equals next index
 * 
 * else iterate over array and return sorted array
 */

// using pointers sort for when you no were things are espicially for a range of numbers

const cyclic_sort = function (nums) {
let index = 0;

while (index < nums.length){
  let y = nums[index] -1; 
  if (y != index){
    let nextIndex = nums[index] - 1;
    let x = nums[index];

    nums[index] = nums[nextIndex];
    nums[nextIndex] = x;
  } else{
    index++;
  }

}
 

  return nums;
};

console.log(`${cyclic_sort([3, 1, 5, 4, 2])}`);
console.log(`${cyclic_sort([2, 6, 4, 3, 1, 5])}`);
console.log(`${cyclic_sort([1, 5, 6, 4, 3, 2])}`);

// Solution
// -----
// function cyclic_sort(nums) {
//   let i = 0;
//   while (i < nums.length) {
//     const j = nums[i] - 1;
    if (nums[i] !== nums[j]) {
//       [nums[i], nums[j]] = [nums[j], nums[i]]; // swap array destructuring
//     } else {
//       i += 1;
//     }
//   }
//   return nums;
// }

// -----

// Time complexity #
// The time complexity of the above algorithm is O(n). Although we are not incrementing the index i when swapping the numbers, this will result in more than ‘n’ iterations of the loop, but in the worst-case scenario, the while loop will swap a total of ‘n-1’ numbers and once a number is at its correct index, we will move on to the next number by incrementing i. So overall, our algorithm will take O(n) + O(n-1) which is asymptotically equivalent to O(n).

// Space complexity #
// The algorithm runs in constant space O(1).

File: mathewDavidov/assignments/cyclicSort/cyclicSortEasy/cyclicSortEasy.java

/**
 * Problem Statement: We are given an array containing ‘n’ objects. Each object,
 * when created, was assigned a unique number from 1 to ‘n’ based on their
 * creation sequence. This means that the object with sequence number ‘3’ was
 * created just before the object with sequence number ‘4’. Write a function to
 * sort the objects in-place on their creation sequence number in O(n) and
 * without any extra space. For simplicity, let’s assume we are passed an
 * integer array containing only the sequence numbers, though each number is
 * actually an object.
 * 
 * Example:
 * Input: [3, 1, 5, 4, 2]
 * Output: [1, 2, 3, 4, 5]
 * 
 * [3, 1, 5, 4, 2] -> swap 3 and 5
 * [5, 1, 3, 4, 2] -> swap 5 and 2
 * [2, 1, 3, 4, 5] -> swap 2 and 1
 * [1, 2, 3, 4, 5] -> increment i while nums[i] == nums[j]
 * end
 * 
 * create a swap method
 * ---
 * initialize iterator to 0
 * while i is less than the length of the array:
 *   intialize j to nums[i] - 1
 *   if nums[i] != nums[j]:
 *     swap the indicies at i and j
 *   otherwise increment i
 * 
 * Runtime:
 * Time: O(n)
 * Space: O(1)
 */
import java.util.*;

public class cyclicSortEasy {
    public static void sort(int[] nums) {
        int i = 0;
        while (i < nums.length) {
            int j = nums[i] - 1;

            if (nums[i] != nums[j]) {
                swap(nums, i, j);
            } else {
                i++;
            }
        }
    }

    public static void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    public static void main(String[] args) {
        int[] arr = new int[] { 3, 1, 5, 4, 2 };
        sort(arr);
        System.out.println(Arrays.toString(arr));

        arr = new int[] { 2, 6, 4, 3, 1, 5 };
        sort(arr);
        System.out.println(Arrays.toString(arr));

        arr = new int[] { 1, 5, 6, 4, 3, 2 };
        sort(arr);
        System.out.println(Arrays.toString(arr));
    }
}
